chain/network/src/peer_manager/network_state/mod.rs

```diff
@@ -34,9 +34,25 @@ pub(crate) const LIMIT_PENDING_PEERS: usize = 60;
 /// We send these messages multiple times to reduce the chance that they are lost
 const IMPORTANT_MESSAGE_RESENT_COUNT: usize = 3;
 
+/// How much time to wait after we send update nonce request before disconnecting.
+/// This number should be large to handle pair of nodes with high latency.
+const WAIT_ON_TRY_UPDATE_NONCE: time::Duration = time::Duration::seconds(6);
+/// If we see an edge between us and other peer, but this peer is not a current connection, wait this
+/// timeout and in case it didn't become a connected peer, broadcast edge removal update.
+const WAIT_PEER_BEFORE_REMOVE: time::Duration = time::Duration::seconds(6);
+
 pub(crate) struct NetworkState {
+    /// Arbiter for executing mutable operation on NetworkState.
+    /// Async methods of NetworkState are not cancellable,
+    /// so calling them from, for example, PeerActor is dangerous because
+    /// PeerActor can be stopped at any moment.
+    /// WARNING: DO NOT spawn infinite futures/background loops on this arbiter,
+    /// as it will be automatically closed only when the NetworkState is dropped.
+    arbiter: actix::ArbiterHandle,
```

Comment:
I've made the drop blocking for the sake of tests. Is that ok?

Reply:
I want the tasks spawned on this arbiter to be non-cancellable. If I use current() the tasks will end up on a random runtime of unknown lifetime.

Reply:
THere's 1-1 correspondence between runtimes and arbiters, so we can us rt from the arbiter. Actix API doesnt' make this easy though: 


```rust
        let handle = Arc::new(OnceCell::new());
        let arbiter = actix::Arbiter::with_tokio_rt({
            let handle = Arc::clone(&handle);
            || {
                let rt = tokio::runtime::Builder::new_current_thread()
                    .enable_io()
                    .enable_time()
                    .build()
                    .unwrap();
                handle.set(rt.handle().clone());
                rt
            }
        });
        let handle = handle.wait().clone();

        Self {
            handle,
```


:thinking: I guess its better to stick with `arbiter` then. It would be cleaner to _not_ depend on actix here, but not at the expense of the above roundabountness..


Actually no, wait, now that we do spawn arbiter here, the code doesnt' really make sense... We basically just use actix to spawn a thread and a runtime, might as well just do that directly, no? 

```diff
diff --git a/chain/network/src/peer_manager/network_state/mod.rs b/chain/network/src/peer_manager/network_state/mod.rs
index d4565655f..50246be7f 100644
--- a/chain/network/src/peer_manager/network_state/mod.rs
+++ b/chain/network/src/peer_manager/network_state/mod.rs
@@ -42,13 +42,10 @@ const WAIT_ON_TRY_UPDATE_NONCE: time::Duration = time::Duration::seconds(6);
 const WAIT_PEER_BEFORE_REMOVE: time::Duration = time::Duration::seconds(6);
 
 pub(crate) struct NetworkState {
-    /// Arbiter for executing mutable operation on NetworkState.
-    /// Async methods of NetworkState are not cancellable,
-    /// so calling them from, for example, PeerActor is dangerous because
-    /// PeerActor can be stopped at any moment.
-    /// WARNING: DO NOT spawn infinite futures/background loops on this arbiter,
-    /// as it will be automatically closed only when the NetworkState is dropped.
-    arbiter: actix::ArbiterHandle,
+    /// Dedicated runtime for `NetworkState` which runs in a separate thread.
+    runtime: tokio::runtime::Handle,
+    notify: Arc<tokio::sync::Notify>,
+    thread: Option<std::thread::JoinHandle<()>>,
     /// PeerManager config.
     pub config: Arc<config::VerifiedConfig>,
     /// When network state has been constructed.
@@ -90,7 +87,8 @@ pub(crate) struct NetworkState {
 
 impl Drop for NetworkState {
     fn drop(&mut self) {
-        self.arbiter.stop();
+        self.notify.notify_one();
+        self.thread.take().unwrap().join().unwrap();
     }
 }
 
@@ -105,8 +103,17 @@ impl NetworkState {
         peer_manager_addr: Recipient<WithSpanContext<PeerToManagerMsg>>,
     ) -> Self {
         let graph = Arc::new(RwLock::new(routing::GraphWithCache::new(config.node_id())));
+        let notify = Arc::new(tokio::sync::Notify::new());
+        let rt = tokio::runtime::Builder::new_current_thread().enable_all().build().unwrap();
+        let handle = rt.handle().clone();
+        let thread = std::thread::spawn({
+            let notify = Arc::clone(&notify);
+            move || rt.block_on(notify.notified())
+        });
         Self {
-            arbiter: actix::Arbiter::new().handle(),
+            runtime: handle,
+            notify,
+            thread: Some(thread),
             routing_table_addr: routing::Actor::spawn(clock.clone(), store.clone(), graph.clone()),
             graph,
             genesis_id,
@@ -298,7 +305,7 @@ impl NetworkState {
         self.routing_table_view.add_local_edges(&edges);
         let this = self.clone();
         let clock = clock.clone();
-        self.arbiter.spawn(async move {
+        self.runtime.spawn(async move {
             match this
                 .routing_table_addr
                 .send(routing::actor::Message::AddVerifiedEdges { edges }.with_span_context())
@@ -368,7 +375,7 @@ impl NetworkState {
                 (Some(conn), EdgeState::Removed) => {
                     let this = self.clone();
                     let conn = conn.clone();
-                    self.arbiter.spawn(async move {
+                    self.runtime.spawn(async move {
                         conn.send_message(Arc::new(PeerMessage::RequestUpdateNonce(
                             PartialEdgeInfo::new(
                                 &node_id,
@@ -394,7 +401,7 @@ impl NetworkState {
                     let this = self.clone();
                     let clock = clock.clone();
                     let other_peer = other_peer.clone();
-                    self.arbiter.spawn(async move {
+                    self.runtime.spawn(async move {
                         // This edge says this is an connected peer, which is currently not in the set of connected peers.
                         // Wait for some time to let the connection begin or broadcast edge removal instead.
                         tokio::time::sleep(WAIT_PEER_BEFORE_REMOVE.try_into().unwrap()).await;
```

Reply:
looks great! I thought about something along these lines, but I thought that
a) "current thread" runtime has to be created on the target thread
b) constructing/destructing a runtime from another runtime is not allowed.

Reply:
actually there are 2 problems:
1. we want the runtime to be shutdown when the actix System is shutting down
2. with the drop implemented like that it might happen (and I see it in tests) that a thread will wait for itself to terminate, which causes a deadlock.